### PR TITLE
Enable CommonSubexpressionElimination in training.

### PR DESCRIPTION
**Pull request by kit1980**

### CODE REVIEW

**Review by MaximKalininMS — DISMISSED** [review]

File: orttraining/orttraining/core/optimizer/graph_transformer_utils.cc [hunk 1]
```diff
@@ -75,6 +76,9 @@ std::vector<std::unique_ptr<GraphTransformer>> GeneratePreTrainingTransformers(
       rule_transformer->Register(make_unique<NonZeroShapeSetter>());
       rule_transformer->Register(make_unique<InsertSoftmaxCrossEntropyLossOutput>());
 
+      // Remove duplicate nodes. Must be applied before any recompute transformations.
+      transformers.emplace_back(onnxruntime::make_unique<CommonSubexpressionElimination>(compatible_eps));
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/optimizer/graph_transformer_utils.cc:80]: I think this still will be in conflict with when we enable recompute. The graph transformer manager will apply the transformations multiple times. 

So CommonSubexpressionElimination will eliminate the recompute nodes that was inserted in the last iteration.

**Comment by kit1980** [comment]
/asp run Windows CPU CI Pipeline (x86_build release)

File: orttraining/orttraining/core/optimizer/graph_transformer_utils.cc [hunk 2]
```diff
@@ -75,6 +76,9 @@ std::vector<std::unique_ptr<GraphTransformer>> GeneratePreTrainingTransformers(
       rule_transformer->Register(make_unique<NonZeroShapeSetter>());
       rule_transformer->Register(make_unique<InsertSoftmaxCrossEntropyLossOutput>());
 
+      // Remove duplicate nodes. Must be applied before any recompute transformations.
+      transformers.emplace_back(onnxruntime::make_unique<CommonSubexpressionEliminationApplyOnce>(compatible_eps));
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/optimizer/graph_transformer_utils.cc:80]: we also need to disable the CommonSubexpressionElimination in inference session. 

Maybe with #ifdef ENABLE_TRAINING ？

**Review by SherlockNoMad — APPROVED** [review]

File: orttraining/orttraining/core/optimizer/graph_transformer_utils.cc (hunk 2 above)
Comment (kit1980) [reply orttraining/orttraining/core/optimizer/graph_transformer_utils.cc:80]: As per offline discussion, submitting this as is, and will figure out recompute interactions later.

**Pull request merged**